spec/consensus/light-client.md

```diff
@@ -1,318 +1,577 @@
+   }
 
-- For each header _h_ it has locally stored, the lite client stores whether
-  it trusts _h_. We write _trust(h) = true_, if this is the case.
+   type SignedHeader struct {
+        Header        Header
+        Commit        Commit            // commit for the given header
+   }
 
-- Validator fields. We will write a validator as a tuple _(v,p)_ such that
-  - _v_ is the identifier (we assume identifiers are unique in each validator set)
-  - _p_ is its voting powers
+   type ValidatorSet struct {
+        Validators         []Validator
+        TotalVotingPower   int64
+   }
 
-### Functions
+   type Validator struct {
+        Address       Address           // validator address (we assume validator's addresses are unique)
+        VotingPower   int64             // validator's voting power
+   }
 
-For the purpose of this lite client specification, we assume that the Tendermint Full Node exposes the following function over Tendermint RPC:
+   type TrustedState {
+        SignedHeader   SignedHeader
+        ValidatorSet   ValidatorSet
+   }
+ ```
 
+### Functions
+
+For the purpose of this lite client specification, we assume that the Tendermint Full Node exposes the following functions over Tendermint RPC:
 ```go
+    // returns signed header: Header with Commit, for the given height
     func Commit(height int64) (SignedHeader, error)
-      // returns signed header: header (with the fields from
-      // above) with Commit that include signatures of
-      // validators that signed the header
-
 
-    type SignedHeader struct {
-      Header        Header
-      Commit        Commit
-    }
+    // returns validator set for the given height
+    func Validators(height int64) (ValidatorSet, error)
 ```
 
-### Definitions
+Furthermore, we assume the following auxiliary functions:
+```go
+    // returns the validator set for the given validator hash
+    func validators(validatorsHash []byte) ValidatorSet
 
-- _tp_: trusting period
-- for realtime _t_, the predicate _correct(v,t)_ is true if the validator _v_
-  follows the protocol until time _t_ (we will see about recovery later).
+    // returns true if commit corresponds to the block data in the header; otherwise false
+    func matchingCommit(header Header, commit Commit) bool
 
-### Tendermint Failure Model
+    // returns the set of validators from the given validator set that committed the block
+    // it does not assume signature verification
+    func signers(commit Commit, validatorSet ValidatorSet) []Validator
 
-If a block _h_ is generated at time _bfttime_ (and this time is stored in the block), then a set of validators that hold more than 2/3 of the voting power in h.Header.NextV is correct until time h.Header.bfttime + tp.
+    // return the voting power the validators in v1 have according to their voting power in set v2
+    // it assumes signature verification so it can be computationally expensive
```

Comment:
This can't be right since it doesn't have access to the commit!

I think probably `signers` is just needed for the model, not for implementation, and votingPowerIn should take a commit and check signatures ?

Reply:
You are right; it's actually the opposite: signers is verifying signatures and voting power is just doing simple math. I left both functions for a time being as it is maybe a bit more intuitive like this. The reason is that we want to get signers of the commit in the corresponding validator set (untrusted validator set), and then compute for those validators voting power in the trusted validator set.